File: packages/x/src/services/premium.ts
// SPDX-License-Identifier: Apache-2.0
import type { XClient } from "../core/client";
import { PREMIUM_GRAPHQL_OPS } from "../config/x-surface";

/** Known upsell surface keys on x.com (from Upsells GraphQL). */
export type UpsellSurfaceKey =
  | "UserProfileName"
  | "UserProfileHeader"
  | "HomeSidebar"
  | "PremiumNav"
  | "HomeNav"
  | string;

/** Checkout SKUs in responsive-web main bundle (2026-06); not separate GraphQL ops. */
export type ProductCategory =
  | "BlueVerified"
  | "BlueVerified3Months"
  | "BlueVerified6Months"
  | "BlueVerifiedPlus"
  | "BlueVerifiedPlus3Months"
  | "BlueVerifiedPlus6Months"
  | "PremiumBasic"
  | string;
export type ChargeInterval = "Month" | "Year" | string;

export interface UpsellDestination {
  charge_interval?: ChargeInterval;
  product_category?: ProductCategory;
}

export interface UpsellConfigEntry {
  key: UpsellSurfaceKey;
  attribution_referrer?: string;
  destination?: UpsellDestination;
  is_hidden?: boolean;
  variant_key?: string;
  action_label?: string;
  primary_label?: string;
}

export interface PremiumUpsells {
  configs: UpsellConfigEntry[];
  raw: unknown;
}

export interface PremiumAccountFlags {
  is_blue_verified?: boolean;
  premium_gifting_eligible?: boolean;
  creator_subscriptions_count?: number;
  super_follow_eligible?: boolean;
  super_followers_count?: number;
  is_super_follow_subscriber?: boolean;
  can_access_payments?: boolean;
}

/** Parse `Upsells` → viewer_v2.upsell_config_for_surfaces. */
export function parseUpsellsResponse(json: unknown): PremiumUpsells {
  const configs: UpsellConfigEntry[] = [];
  const surfaces =
    (json as any)?.data?.viewer_v2?.user_results?.result?.upsell_config_for_surfaces?.configs;
  if (!Array.isArray(surfaces)) {
    return { configs, raw: json };
  }
  for (const item of surfaces) {
    const key = item?.key as string;
    const val = item?.value;
    const def = val?.default_content;
    const variant = val?.variant_config?.variants?.[0];
    const content = variant ?? def;
    configs.push({
      key,
      attribution_referrer: val?.attribution_params?.referrer,
      destination: content?.destination,
      is_hidden: def?.is_hidden,
      variant_key: content?.key,
      action_label:
        content?.render_properties?.action_label ??
        content?.render_properties?.cta?.action_label,
      primary_label: content?.render_properties?.primary_label,
    });
  }
  return { configs, raw: json };
}

/** Extract premium-related flags from Viewer + UserByScreenName payloads. */
export function parsePremiumFlags(viewerJson: unknown, userJson?: unknown): PremiumAccountFlags {
  const user =
    (viewerJson as any)?.data?.viewer?.user_results?.result ??
    (userJson as any)?.data?.user?.result;
  const viewerRoot = (viewerJson as any)?.data;
  return {
    is_blue_verified: user?.is_blue_verified,
    premium_gifting_eligible: user?.premium_gifting_eligible,
    creator_subscriptions_count: user?.creator_subscriptions_count,
    super_follow_eligible: user?.super_follow_eligible,
    super_followers_count: viewerRoot?.viewer?.super_followers_count,
    is_super_follow_subscriber: viewerRoot?.is_super_follow_subscriber,
    can_access_payments: viewerRoot?.can_access_payments,
  };
}

export async function fetchPremiumUpsells(client: XClient): Promise<PremiumUpsells> {
  const json = await client.graphqlWaiting("Upsells", {});
  return parseUpsellsResponse(json);
}

export async function fetchPremiumBundle(
  client: XClient,
  handle: string,
  userId: string,
): Promise<{ upsells: PremiumUpsells; flags: PremiumAccountFlags; raw: Record<string, unknown> }> {
  const [upsellsJson, viewerJson, userJson] = await Promise.all([
    client.graphqlWaiting("Upsells", {}),
    client.graphqlWaiting("Viewer", { withCommunitiesMemberships: true }),
    client.graphqlWaiting("UserByScreenName", {
      screen_name: handle,
      withSafetyModeUserFields: true,
    }),
  ]);
  return {
    upsells: parseUpsellsResponse(upsellsJson),
    flags: parsePremiumFlags(viewerJson, userJson),
    raw: { Upsells: upsellsJson, Viewer: viewerJson, UserByScreenName: userJson },
  };
}

/** Variable templates for full Premium GraphQL coverage (see scripts/x-premium-dump.ts). */
export function premiumGraphqlVariables(
  handle: string,
  userId: string,
): Record<string, Record<string, unknown>> {
  return {
    Upsells: {},
    Viewer: { withCommunitiesMemberships: true },
    UserByScreenName: { screen_name: handle, withSafetyModeUserFields: true },
    UserCreatorSubscriptions: { userId, count: 50, includePromotedContent: false },
    CreatorSubscriptionsTimeline: { userId, count: 50, includePromotedContent: false },
    UserCreatorSubscribers: { userId, count: 50, includePromotedContent: false },
    SuperFollowers: { userId, count: 50, includePromotedContent: false },
    BlueVerifiedFollowers: { userId, count: 20, includePromotedContent: false },
    UserArticlesTweets: { userId, count: 20, includePromotedContent: true, withVoice: true },
    UsersVerifiedAvatars: { userIds: [userId] },
    EnableVerifiedPhoneLabel: {},
    DisableVerifiedPhoneLabel: {},
  };
}

/** Fetch every op in PREMIUM_GRAPHQL_OPS; failures are recorded, not thrown. */
export async function fetchAllPremiumGraphql(
  client: XClient,
  handle: string,
  userId: string,
): Promise<Record<string, { ok: true; data: unknown } | { ok: false; error: string }>> {
  const vars = premiumGraphqlVariables(handle, userId);
  const out: Record<string, { ok: true; data: unknown } | { ok: false; error: string }> = {};
  for (const op of PREMIUM_GRAPHQL_OPS) {
    try {
      const data = await client.graphqlWaiting(op, vars[op] ?? {});
      out[op] = { ok: true, data };
    } catch (err: unknown) {
      out[op] = { ok: false, error: err instanceof Error ? err.message : String(err) };
    }
  }
  return out;
}

File: rust-bridge/crates/x-client/src/premium.rs
// SPDX-License-Identifier: Apache-2.0
//! X Premium / Blue Verified — Upsells GraphQL and account flags.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use std::time::Duration;

use crate::client::XClient;
use crate::Result;

const PREMIUM_MAX_WAIT: Duration = Duration::from_secs(120);

/// Known upsell surface keys on x.com.
pub const UPSELL_SURFACES: &[&str] = &[
    "UserProfileName",
    "UserProfileHeader",
    "HomeSidebar",
    "PremiumNav",
    "HomeNav",
];

pub use crate::surface::{PREMIUM_GRAPHQL_OPS, PRODUCT_SKUS};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PremiumAccountFlags {
    #[serde(default)]
    pub is_blue_verified: Option<bool>,
    #[serde(default)]
    pub premium_gifting_eligible: Option<bool>,
    #[serde(default)]
    pub creator_subscriptions_count: Option<u64>,
    #[serde(default)]
    pub super_follow_eligible: Option<bool>,
    #[serde(default)]
    pub super_followers_count: Option<u64>,
    #[serde(default)]
    pub is_super_follow_subscriber: Option<bool>,
    #[serde(default)]
    pub can_access_payments: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpsellConfigEntry {
    pub key: String,
    #[serde(default)]
    pub product_category: Option<String>,
    #[serde(default)]
    pub charge_interval: Option<String>,
    #[serde(default)]
    pub is_hidden: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PremiumUpsells {
    pub configs: Vec<UpsellConfigEntry>,
}

fn dig<'a>(v: &'a Value, path: &[&str]) -> Option<&'a Value> {
    let mut cur = v;
    for key in path {
        cur = cur.get(key)?;
    }
    Some(cur)
}

/// Parse `Upsells` → `viewer_v2.upsell_config_for_surfaces`.
pub fn parse_upsells_response(json: &Value) -> PremiumUpsells {
    let mut configs = Vec::new();
    let Some(arr) = dig(
        json,
        &[
            "data",
            "viewer_v2",
            "user_results",
            "result",
            "upsell_config_for_surfaces",
            "configs",
        ],
    )
    .and_then(|v| v.as_array())
    else {
        return PremiumUpsells { configs };
    };

    for item in arr {
        let key = item.get("key").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let val = item.get("value");
        let def = val.and_then(|v| v.get("default_content"));
        let variant = val
            .and_then(|v| v.get("variant_config"))
            .and_then(|v| v.get("variants"))
            .and_then(|v| v.as_array())
            .and_then(|a| a.first());
        let content = variant.or(def);
        let destination = content.and_then(|c| c.get("destination"));
        configs.push(UpsellConfigEntry {
            key,
            product_category: destination
                .and_then(|d| d.get("product_category"))
                .and_then(|v| v.as_str())
                .map(str::to_owned),
            charge_interval: destination
                .and_then(|d| d.get("charge_interval"))
                .and_then(|v| v.as_str())
                .map(str::to_owned),
            is_hidden: def
                .and_then(|d| d.get("is_hidden"))
                .and_then(|v| v.as_bool()),
        });
    }
    PremiumUpsells { configs }
}

/// Extract premium-related flags from Viewer + optional UserByScreenName JSON.
pub fn parse_premium_flags(viewer: &Value, user_by_screen: Option<&Value>) -> PremiumAccountFlags {
    let user = viewer
        .pointer("/data/viewer/user_results/result")
        .or_else(|| user_by_screen.and_then(|u| u.pointer("/data/user/result")));
    let viewer_root = viewer.get("data");
    PremiumAccountFlags {
        is_blue_verified: user
            .and_then(|u| u.get("is_blue_verified"))
            .and_then(|v| v.as_bool()),
        premium_gifting_eligible: user
            .and_then(|u| u.get("premium_gifting_eligible"))
            .and_then(|v| v.as_bool()),
        creator_subscriptions_count: user
            .and_then(|u| u.get("creator_subscriptions_count"))
            .and_then(|v| v.as_u64()),
        super_follow_eligible: user
            .and_then(|u| u.get("super_follow_eligible"))
            .and_then(|v| v.as_bool()),
        super_followers_count: viewer_root
            .and_then(|r| r.pointer("/viewer/super_followers_count"))
            .and_then(|v| v.as_u64()),
        is_super_follow_subscriber: viewer_root
            .and_then(|r| r.get("is_super_follow_subscriber"))
            .and_then(|v| v.as_bool()),
        can_access_payments: viewer_root
            .and_then(|r| r.get("can_access_payments"))
            .and_then(|v| v.as_bool()),
    }
}

impl XClient {
    /// Fetch parsed Upsells for the logged-in viewer.
    pub async fn premium_upsells(&self) -> Result<PremiumUpsells> {
        let json = self
            .graphql_waiting("Upsells", serde_json::json!({}), None, PREMIUM_MAX_WAIT)
            .await?;
        Ok(parse_upsells_response(&json))
    }

    /// Upsells + Viewer + UserByScreenName in parallel.
    pub async fn premium_bundle(
        &self,
        handle: &str,
    ) -> Result<(PremiumUpsells, PremiumAccountFlags, Value, Value, Value)> {
        let (upsells, viewer, user) = tokio::join!(
            self.graphql_waiting("Upsells", serde_json::json!({}), None, PREMIUM_MAX_WAIT),
            self.graphql_waiting(
                "Viewer",
                serde_json::json!({ "withCommunitiesMemberships": true }),
                None,
                PREMIUM_MAX_WAIT
            ),
            self.graphql_waiting(
                "UserByScreenName",
                serde_json::json!({
                    "screen_name": handle,
                    "withSafetyModeUserFields": true
                }),
                None,
                PREMIUM_MAX_WAIT
            ),
        );
        let upsells = upsells?;
        let viewer = viewer?;
        let user = user?;
        let parsed = parse_upsells_response(&upsells);
        let flags = parse_premium_flags(&viewer, Some(&user));
        Ok((parsed, flags, upsells, viewer, user))
    }
}